packages/superset-ui-core/src/query/normalizeOrderBy.ts

```diff
@@ -19,46 +19,55 @@
 import isEmpty from 'lodash/isEmpty';
 import isBoolean from 'lodash/isBoolean';
 
-import { QueryObject } from './types';
+import { FormDataResidual, QueryFormOrderBy } from './types';
+import { t } from '../translation';
+import { ensureIsArray } from '../utils';
 
-export default function normalizeOrderBy(queryObject: QueryObject): QueryObject {
-  if (Array.isArray(queryObject.orderby) && queryObject.orderby.length > 0) {
-    // ensure a valid orderby clause
-    const orderbyClause = queryObject.orderby[0];
-    if (
+const defaultOrderby = undefined;
+
+export default function normalizeOrderBy(
+  formData: FormDataResidual,
+): QueryFormOrderBy[] | typeof defaultOrderby {
+  const orderbyClauses = ensureIsArray(formData.orderby).map(item => {
+    // value can be in the format of `['["col1", true]', '["col2", false]']`,
+    // where the option strings come directly from `order_by_choices`.
+    if (typeof item === 'string') {
+      try {
+        return JSON.parse(item);
+      } catch (error) {
+        throw new Error(t('Found invalid orderby options'));
+      }
+    }
+    return item;
+  });
+  // ensure a valid orderby clause
+  const validatedOrderbys = orderbyClauses.filter(
+    orderbyClause =>
       Array.isArray(orderbyClause) &&
       orderbyClause.length === 2 &&
       !isEmpty(orderbyClause[0]) &&
-      isBoolean(orderbyClause[1])
-    ) {
-      return queryObject;
-    }
+      isBoolean(orderbyClause[1]),
+  );
+  if (validatedOrderbys.length > 0) {
+    return validatedOrderbys;
   }
 
-  // ensure that remove invalid orderby clause
-  const cloneQueryObject = { ...queryObject };
-  delete cloneQueryObject.timeseries_limit_metric;
-  delete cloneQueryObject.order_desc;
-  delete cloneQueryObject.orderby;
-
-  const isAsc = !queryObject.order_desc;
+  // the `timeseries_limit_metric` and `order_desc` are used for orderby
+  const isAsc = !formData.order_desc;
```

Comment:
I personally think we should avoid using `timeseries_limit_metric` to order the main query. I know many charts, especially legacy ones, seem to use it both for series limiting (subquery) and ordering the main query, but IMO timeseries charts `is_timeseries = true` should rather default to sorting primarily by the temporal column, and only then default to the `timeseries_limit_metric` if necessary.

A good example of this is the table chart, where setting the 'Include time' control to `true` orders the data in descending order according to the metric, NOT the temporal column. This has a very weird effect, where the data is displayed in non-chronological order.

Reply:
Let's discuss offline how to deal with the `order by` logic of inner and outer queries. ^_^

Reply:
okay. I will remove the transformation.